### DISCUSSION TITLE
The best way to return message from external environment to wasm code

### DISCUSSION THREAD

**Discussion by changyc9928** [wasmerio/wasmer#2879] [Q&A]
Hi,
I'm stuck on returning data from the Rust VM to the wasm code since Rust can only return an integer to the wasm code. So, I'm writing here to ask whether there is any good solution to allow me to return any complex data from an external environment back to the wasm code when the exported function is called in the VM?

Thank you.

**Comment by jcaesar** [comment]
You'll need to have the WASM code pass a pointer (and length) to the host, and have the host write to the wasm memory. You can either do that [manually](https://github.com/wasmerio/wasmer/blob/master/examples/exports_memory.rs#L94), or e.g. use wit-bindgen.
